src/Misc/PerformanceViewer/performanceViewerCollector.ts

```diff
@@ -0,0 +1,225 @@
+        const numberOfIndices = this.datasets.startingIndices.itemLength;
+        let startingIndex = 0;
+
+        if (numberOfIndices > 0) {
+            const previousStartingIndex = this.datasets.startingIndices.at(numberOfIndices - 1);
+            startingIndex = previousStartingIndex + this.datasets.data.at(previousStartingIndex + PerformanceViewerCollector.NumberOfPointsOffset) + PerformanceViewerCollector.SliceDataOffset;
+        }
+
+        this.datasets.startingIndices.push(startingIndex);
+
+        // add the first 2 items in our slice.
+        this.datasets.data.push(timestamp);
+        this.datasets.data.push(numPoints);
+
+        // add the values inside the slice.
+        this.datasets.ids.forEach((id: string) => {
+            const strategy = this._strategies.get(id);
+
+            if (!strategy) {
+                return;
+            }
+
+            this.datasets.data.push(strategy.getData());
+        });
+
+        if (this.datasetObservable.hasObservers()) {
+            const slice: number[] = [timestamp, numPoints];
+
+            for (let i = 0; i < numPoints; i++) {
+                slice.push(this.datasets.data.at(startingIndex + PerformanceViewerCollector.SliceDataOffset + i));
+            }
+
+            this.datasetObservable.notifyObservers(slice);
+        }
+    }
+
+    /**
+     * Collects and then sends the latest slice to any observers by using the appropriate strategy when the user wants.
+     * This method does not add onto the collected data accessible via the datasets variable.
+     */
+    public getCurrentSlice() {
+        const timestamp = PrecisionDate.Now - this._startingTimestamp;
+        const numPoints = this.datasets.ids.length;
+        const slice: number[] = [timestamp, numPoints];
+
+        // add the values inside the slice.
+        this.datasets.ids.forEach((id: string) => {
+            const strategy = this._strategies.get(id);
+
+            if (!strategy) {
+                return;
+            }
+
+            if (this.datasetObservable.hasObservers()) {
+                slice.push(strategy.getData());
+            }
+        });
+
+        if (this.datasetObservable.hasObservers()) {
+            this.datasetObservable.notifyObservers(slice);
+        }
+
+    }
+
+    /**
+     * Updates a property for a dataset's metadata with the value provided.
+     * @param id the id of the dataset which needs its metadata updated.
+     * @param prop the property to update.
+     * @param value the value to update the property with.
+     */
+    public updateMetadata<T extends keyof IPerfMetadata>(id: string, prop: T, value: IPerfMetadata[T]) {
+        const meta = this._datasetMeta.get(id);
+
+        if (!meta) {
+            return;
+        }
+
+        meta[prop] = value;
+
+        this.metadataObservable.notifyObservers(this._datasetMeta);
+    }
+
+    /**
+     * Starts the realtime collection of data.
+     */
+    public start() {

```

Comment:
If the user calls start, stop and then start again it will have a huge gap in timestamps. Might be what the user wanted, but you could add a flag parameter to reset the datasets (maybe the default behavior?).
Maybe a future improvement would be to set an ID to this dataset, so every call to start it will save a new dataset so that the user can go back to older datasets but still get a new one when start is called again

Reply:
We now reset the indices and the data.